### PR TITLE
Speedup segmented sort with large nsort

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #77188

Follow up to gh-77100

Instead of calling `at::arange`, I repurpose the existing kernels to
achieve the same effect. I've also fixed the 2d case bug where
the pointer was advanced by `n` which equals `nsegment * nsort` after
only processing `nsort` elements.

### CODE REVIEW

**Comment by ngimel** [comment]
I reverted that PR because of nsort bug already :-)

**Comment by peterbell10** [comment]
I already rebased to after the revert before submitting the PR.

**Comment by ngimel** [comment]
Can you also please add a test that I added to the reverted PR?

File: aten/src/ATen/native/cuda/Sort.cu [hunk 1]
```diff
@@ -241,13 +243,38 @@ __global__ void fill_index_and_segment_kernel(
...
+  using namespace at::cuda::detail;
+  auto allocator = at::cuda::getCUDADeviceAllocator();
+  auto stream = at::cuda::getCurrentCUDAStream();
+  dim3 block = CUDA_NUM_THREADS;
+  dim3 grid = GET_BLOCKS(nsort);
+  c10::DeviceArray<int64_t> indices(*allocator, nsort);
+  at::cuda::detail::IntDivider<uint32_t> nsort_divider(nsort);
+
+  for (auto i: c10::irange(nsegments)){
+    fill_reverse_indices_kernel<<<grid, block, 0, stream>>>(
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/Sort.cu:268]: oh, my mistake - we don't need to refill indices every time as cub doesn't modify them.

**Review by ngimel — APPROVED** [review]
This looks good, but I can't merge now as github is rate limited and CI is not running

**Pull request closed without merging**